content/blog/posts/vectors-matrices-matrix-order.md

```diff
@@ -0,0 +1,732 @@
+   {
+      // non-contiguous element access
+      std::cout << rowMajor[r][c] << " ";
+   }
+   
+   std::cout << "\n";
+}
+```
+
+In the extreme example above, each row is a L1 cache line in length, so accessing each element using column-major indexing is not only non-contiguous memory access but a worst case scenario where every element access results in a cache miss. However, this is only because we are implicitly using the language’s assumption to store the matrix data but using our own assumption about matrix order to access the data. 
+
+As we have seen in the previous section, there is nothing stopping us from storing our matrices using a column-major layout if we wish to access our elements column-by-column rather than row-by-row so ideally we would pick conventions that match closest to our expected element access patterns. For transformation matrices, this question has a relatively straightforward answer: as our matrices will contain the basis vectors for our vector transformations, when we pick the matrix order convention that matches our vector convention, the basis vectors in our transformation matrices will be guaranteed to have contiguous components. On the flipside, if we don’t match our matrix order and vector conventions, our basis vector components will be separated in memory by a stride matching the matrix dimension of the matrix order  convention used to store the matrix. Consider the following example of retrieving the x-axis basis vector of a 2d rotation matrix for column-vectors stored in row-major:
+
+```c++
+
+// Rotation of approx. 90 degrees (in radians)
+const float theta = 0.707;
+
+// 2x2 matrix in row-major order
+const std::size_t rows = 2;
+const std::size_t cols = 2;
+using row_type = std::array<float, cols>;
+using col_type = std::array<float, rows>;
+using mat2x2_type = std::array<row_type, rows>;
+
+// 2d rotation matrix for column vectors
+const mat2x2_type rowMajor = 
+{
+    {cos(theta), -sin(theta)}, // row 0 
+    {sin(theta),  cos(theta)}  // row 1
+};
+
+// returns the x basis vector
+col_type basisX()
+{
+     // x-axis basis vector components stored non-contiguously
+    return col_type 
+    {
+        rowMajor[0][0],
+        rowMajor[1][0]
+    };
+}
+```
+
+Notice that we have to construct an intermediary array and populate it with the basis vector components due to the fact that they’re not stored contiguously in the array representation of the matrix. Now compare that to the following example of retrieving the x-axis basis vector of the same matrix stored in column-major:
+
+```c++
+// Rotation of approx. 90 degrees (in radians)
+const float theta = 0.707;
+
+// 2x2 matrix in column-major order
+const std::size_t rows = 2;
+const std::size_t cols = 2;
+using col_type = std::array<float, rows>;
+using mat2x2_type = std::array<col_type, cols>;
+
+// 2d rotation matrix for column vectors
+const mat2x2_type columnMajor = 
+{ 
+    {Sx * cos(theta), sin(theta)}, // column 0 
```

Comment:
Should this Sx be here? It hasn't been declared above and I don't think it's necessary for this example.

Reply:
```suggestion
    {cos(theta), sin(theta)}, // column 0 
```